rs/embedders/benches/heap.rs

```diff
@@ -72,57 +90,69 @@ fn throughput(size: Size, step: Step) -> Option<Throughput> {
 fn setup_action(src: Src) -> SetupAction {
     match src {
         Src::Checkpoint => SetupAction::PerformCheckpoint,
+        Src::Mix => SetupAction::PerformCheckpointCallSetup,
         Src::PageDelta | Src::NewAllocation => SetupAction::None,
     }
 }
 
-fn heap_op(op: Op) -> String {
-    match op {
-        Op::Read => "(drop (i64.load (local.get $address)))".into(),
-        Op::Write => "(i64.store (local.get $address) (global.get $counter))".into(),
+fn heap_op(op: Op, mem: Mem) -> String {
+    match (op, mem) {
+        (Op::Read, Mem::Wasm32) => {
+            "(global.set $data (i64.load8_u (i32.wrap_i64 (local.get $address))))"
+        }
+        (Op::Read, Mem::Wasm64) => "(global.set $data (i64.load8_u (local.get $address)))",
+        (Op::Write, Mem::Wasm32) => {
+            "(i64.store8 (i32.wrap_i64 (local.get $address)) (global.get $counter))"
+        }
+        (Op::Write, Mem::Wasm64) => "(i64.store8 (local.get $address) (global.get $counter))",
     }
+    .into()
 }
 
 fn heap_func_init_body(mem: Mem, size: Size, src: Src) -> String {
     let wasm_pages = (size as usize).div_ceil(WASM_PAGE_SIZE_IN_BYTES);
     match src {
-        Src::Checkpoint | Src::PageDelta => String::new(),
+        Src::Checkpoint | Src::PageDelta | Src::Mix => String::new(),
         Src::NewAllocation => format!("(drop (memory.grow (i{mem}.const {wasm_pages})))"),
     }
 }
 
-fn loop_body(op: &str, mem: Mem, dir: Dir, size: Size, step: Step) -> String {
-    let size = size as usize;
+fn loop_body(op: &str, dir: Dir, initial_offset: usize, size: Size, step: Step) -> String {
     let step = step as usize;
+    let end = size as usize - 1;
     match dir {
-        Dir::Forward => format!(
-            r#"
-                (local.set $address (i{mem}.const 0))
+        Dir::Fwd => {
+            format!(
+                r#"
+                (local.set $address (i64.const {initial_offset}))
                 (loop $loop
                     {op}
-                    (local.set $address (i{mem}.add (local.get $address) (i{mem}.const {step})))
-                    (br_if $loop (i{mem}.lt_u (local.get $address) (i{mem}.const {size})))
+                    (local.set $address (i64.add (local.get $address) (i64.const {step})))
+                    (br_if $loop (i64.le_s (local.get $address) (i64.const {end})))
                 )
             "#
-        ),
-        Dir::Backward => format!(
-            r#"
-                (local.set $address (i{mem}.const {size}))
+            )
+        }
+        Dir::Bwd => {
+            format!(
+                r#"
+                (local.set $address (i64.const {end}))
                 (loop $loop
-                    (local.set $address (i{mem}.sub (local.get $address) (i{mem}.const {step})))
                     {op}
```

Comment:
The first iteration is now executed at `size - 1` instead of `size - step`.

Reply:
I believe that's the correct behavior. Previously, we were skipping a few last pages.

Reply:
In the forward direction, we iterate through initial_offset, initial_offset + step, initial_offset + 2 * step, ..., size - step (assuming alignment: size = initial_offset + n * step) so I'd expect the backward direction starts at size - step.

Reply:
Sorry, I confused directions. Thanks. I've fixed that for backward direction it starts at `size - offset - 1`, which is probably what we actually want to make it symmetrical for the forward direction.

Reply:
To me it is even more confusing now: https://github.com/dfinity/ic/pull/4985/files/e978c1d4750e0f3106c61227b0276a1628444fa7..a2996696b38638c7356ee74634af232c02e1e1e3#r2095059600